### PR TITLE
New cookie implementation for the HTTP stack

### PR DESCRIPTION
**Pull request by burzum**
## New prototype implementation for cookies

It needs more tests and I'm not yet happy with the handling of the cookie data and encryption. But I would already like to get feedback on this to continue to develop it into a direction we all agree with.

The code is partially based on https://github.com/hansott/psr7-cookies/ but was changed to fit into our current API and the frameworks current php version. I've had a discussion with Mark Story if we'll directly use it or not. The final decision for the named reasons was to come up with our own implementation. I've also changed the implementation in some aspects, it's not a copy and paste job. I need some advice if we need to give credits and copyright and if yes, how to do it the right way.

Cookies can now be used from wherever the request object and response objects are available.

## Overview of the new implementation

* Uses the request and response objects
* Uses a cookie object to built and modify a cookie
* Generates cookie objects from the request object
* Writes cookies to the response object
* Uses getters and setters
* No longer a component bound controller layer only solution (This change is required to get the cookie authenticator in the new authentication implementation working)
* **TBD**: If possible a BC compatible cookie component that uses the new implementation

## Related issues

* https://github.com/cakephp/authentication/issues/26
* https://github.com/cakephp/cakephp/pull/10203

## History

Former PR from my fork https://github.com/cakephp/cakephp/pull/10203

Created a new PR based on the cake repo and not my fork so other people can contribute to the branch as well without being a member of my fork.

### CODE REVIEW

File: src/Http/Cookie/CookieCryptTrait.php [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+     * @return string Encoded values
+     */
+    protected function _encrypt($value)
+    {
+        if (is_array($value)) {
+            $value = $this->_flatten($value);
+        }
+
+        $encrypt = $this->encryptionCipher;
+        $prefix = 'Q2FrZQ==.';
```
Comment (inoas) [inline src/Http/Cookie/CookieCryptTrait.php]: Can this be explainrd in the doc block?

File: src/Http/Cookie/CookieCryptTrait.php [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+     * @param string $value The value to decode & decrypt.
+     * @return string|array Decoded values.
+     */
+    protected function _decode($value)
+    {
+        if (!$this->isEncryptionEnabled()) {
+            return $this->_expand($value);
+        }
+
+        $prefix = 'Q2FrZQ==.';
```
Comment (inoas) [inline src/Http/Cookie/CookieCryptTrait.php]: Same as above.

**Comment by markstory** [comment]
>  I've also changed the implementation in some aspects, it's not a copy and paste job. I need some advice if we need to give credits and copyright and if yes, how to do it the right way.

What I've seen done in other projects is a small mention in the file header that the implementation was inspired by X.

File: src/Http/Cookie/Cookie.php [hunk 3]
```diff
@@ -0,0 +1,531 @@
...
+     * Constructor
+     *
+     * The constructors args are similar to the native php setcookie() method.
+     * The only difference is the 3rd argument which excepts null or an object
+     * implementing \DateTimeInterface instead an integer.
+     *
+     * @link http://php.net/manual/en/function.setcookie.php
+     * @param string $name Cookie name
+     * @param string|array $value Value of the cookie
+     * @param \DateTimeInterface|null $expiresAt Expiration time and date
```
Comment (markstory) [inline src/Http/Cookie/Cookie.php]: Should this argument use a typehint?

File: src/Http/Cookie/Cookie.php [hunk 4]
```diff
@@ -0,0 +1,531 @@
...
+    }
+
+    /**
+     * Returns a header value as string
+     *
+     * @return string
+     */
+    public function toHeaderValue()
+    {
+        $headerValue[] = sprintf('%s=%s', $this->name, urlencode($this->value));
```
Comment (markstory) [inline src/Http/Cookie/Cookie.php]: Does the name need to be urlencoded as well?

File: src/Http/Cookie/CookieCollection.php [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+namespace Cake\Http\Cookie;
+
+use ArrayIterator;
+use InvalidArgumentException;
+use IteratorAggregate;
+
+/**
+ * Cookie Collection
+ */
+class CookieCollection implements IteratorAggregate
```
Comment (markstory) [inline src/Http/Cookie/CookieCollection.php]: Is there any opportunity to repurpose/extend/reuse the `CookieCollection` in `Cake\Http\Client`?

File: src/Http/Cookie/CookieCryptTrait.php [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+     * @param string|array $value Value to encrypt
+     * @return string Encoded values
+     */
+    protected function _encrypt($value)
+    {
+        if (is_array($value)) {
+            $value = $this->_flatten($value);
+        }
+
+        $encrypt = $this->encryptionCipher;
```
Comment (markstory) [inline src/Http/Cookie/CookieCryptTrait.php]: If this value is `false` should you return early?

File: src/Http/Cookie/RequestCookies.php [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @since         3.5.0
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Http\Cookie;
+
+use InvalidArgumentException;
+use Psr\Http\Message\ServerRequestInterface;
+
+class RequestCookies extends CookieCollection
```
Comment (markstory) [inline src/Http/Cookie/RequestCookies.php]: This is another opportunity to make the client better/reuse existing `Http\Client` code.

File: tests/TestCase/Http/Cookie/CookieCollectionTest.php [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+     */
+    public function testConstructorWithCookieArray()
+    {
+        $cookies = [
+            new Cookie('one', 'one'),
+            new Cookie('two', 'two')
+        ];
+
+        $collection = new CookieCollection($cookies);
+        $this->assertCount(2, $collection);
```
Comment (markstory) [inline tests/TestCase/Http/Cookie/CookieCollectionTest.php]: What about a test for iterating the collection?

File: src/Http/Cookie/CookieInterface.php [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @since         3.5.0
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Http\Cookie;
+
+/**
+ * Cookie Interface
+ */
+interface CookieInterface
```
Comment (robertpustulka) [inline src/Http/Cookie/CookieInterface.php]: Would it make sense to add `getName()` and `getValue()` to the interface?

File: src/Http/Cookie/Cookie.php [hunk 10]
```diff
@@ -0,0 +1,531 @@
...
+            );
+        }
+
+        if (empty($name)) {
+            throw new InvalidArgumentException('The cookie name cannot be empty.');
+        }
+    }
+
+    /**
+     * Gets the raw cookie value
```
Comment (robertpustulka) [inline src/Http/Cookie/Cookie.php]: After calling `expand()` this is no longer "raw" value?
Also this method may return `array` in that case, am I right?

**Comment by robertpustulka** [comment]
AFAIR when we had a first go with reimplementing cookies (before PSR7) there was a question where cookie encryption/decryption would take place. Have you thought about it yet?

Another thing is how to add cookies to the response. Currently the only way is to instantiate `RequestCookies` collection with existing cookies and call `addToResponse()`. Where should this be done? In a middleware?

File: src/Http/Cookie/CookieCollection.php (hunk 5 above)
Comment (markstory) [inline src/Http/Cookie/CookieCollection.php]: @burzum After looking at the various cookie collection classes, I think it is a big enough change that I can do after this merges. I'll open an RFC ticket first so we can align on the changes I have in mind.

**Comment by markstory** [comment]
@burzum Right now there are no methods on `Response` or `ServerRequest` objects that interact with these cookies. It might be a good idea to consider adding methods that work with these new cookies or they could end up feeling pretty clunky to users. This integration can be done in a separate pull request though.

File: src/Http/Cookie/CookieCryptTrait.php (hunk 6 above)
Comment (burzum) [inline src/Http/Cookie/CookieCryptTrait.php]: Maybe better throw an error that you try to encrypt but haven't set a ciper than return?

File: src/Http/Cookie/CookieCollection.php (hunk 5 above)
Comment (burzum) [inline src/Http/Cookie/CookieCollection.php]: We want type save method calls for the future right? So either we use int or a datetime object but not allow both. That's why I've chosen the object.

File: src/Http/Cookie/CookieInterface.php (hunk 9 above)
Comment (burzum) [inline src/Http/Cookie/CookieInterface.php]: It should have setters as well then.

**Comment by markstory** [comment]
@burzum I'll get started on refactoring the various CookieCollections. Would you be able to look at a more ergonomic integration between the new cookie objects and the ServerRequest/Response objects?

**Comment by burzum** [comment]
@markstory I'll talk to you tomorrow or on Tuesday. I'm going to have a busy week due to some appointments. But yes, I'll see what I can do and when I can do it. Just not very much in the upcoming week.

**Pull request merged**